### PR TITLE
[SPARK-29142][PYTHON][ML] Pyspark clustering models support column setters/getters/predict

### PR DESCRIPTION
**Pull request by huaxingao**
### What changes were proposed in this pull request?
Add the following Params classes in Pyspark clustering
```GaussianMixtureParams```
```KMeansParams```
```BisectingKMeansParams```
```LDAParams```
```PowerIterationClusteringParams```

### Why are the changes needed?
To be consistent with scala side

### Does this PR introduce any user-facing change?
Yes. Add the following changes:
```
GaussianMixtureModel
- get/setMaxIter
- get/setFeaturesCol
- get/setSeed
- get/setPredictionCol
- get/setProbabilityCol
- get/setTol
- predict
```

```
KMeansModel
- get/setMaxIter
- get/setFeaturesCol
- get/setSeed
- get/setPredictionCol
- get/setDistanceMeasure
- get/setTol
- predict
```

```
BisectingKMeansModel
- get/setMaxIter
- get/setFeaturesCol
- get/setSeed
- get/setPredictionCol
- get/setDistanceMeasure
- predict
```

```
LDAModel(HasMaxIter, HasFeaturesCol, HasSeed, HasCheckpointInterval):
- get/setMaxIter
- get/setFeaturesCol
- get/setSeed
- get/setCheckpointInterval
```

### How was this patch tested?
Add doctests

### CODE REVIEW

File: python/pyspark/ml/clustering.py [hunk 1]
```diff
@@ -135,10 +156,16 @@ def summary(self):
             raise RuntimeError("No training summary available for this %s" %
                                self.__class__.__name__)
 
+    @since("3.0.0")
+    def predict(self, value):
+        """
+        Predict label for the given features.
+        """
+        return self._call_java("predict", value)
+
```
Comment (zhengruifeng) [inline python/pyspark/ml/clustering.py]: we should add `def predictProbability` as well

File: python/pyspark/ml/clustering.py [hunk 2]
```diff
@@ -347,14 +418,20 @@ class KMeans(JavaEstimator, HasDistanceMeasure, HasFeaturesCol, HasPredictionCol
     >>> df = spark.createDataFrame(data, ["features"])
     >>> kmeans = KMeans(k=2, seed=1)
     >>> model = kmeans.fit(df)
+    >>> model.getDistanceMeasure()
+    'euclidean'
+    >>> model.setPredictionCol("newPrediction")
+    KMeans...
```
Comment (zhengruifeng) [inline python/pyspark/ml/clustering.py]: KMeansModel...

Comment (huaxingao) [reply python/pyspark/ml/clustering.py]: It is ```KMeans_3487dfaa7c0e```

**Review by srowen — COMMENTED** [review]
All this does is refactor the getters, right? the description says it adds setters, but I'm not seeing that at a first look.

Also it's a tricky question, what to do about `@since` annotations when subclasses add them at different times. I might suggest making it the highest version of any of the methods that are removed in favor of a new one.

**Comment by huaxingao** [comment]
@srowen 
This PR also adds the setters. 
Use ```GaussianMixtureModel``` as an example:
before the PR: 
```
class GaussianMixtureModel(JavaModel, JavaMLWritable, JavaMLReadable, HasTrainingSummary):
```
after the PR: 

```
class GaussianMixtureParams(HasMaxIter, HasFeaturesCol, HasSeed, HasPredictionCol,
                            HasProbabilityCol, HasTol):
class GaussianMixtureModel(JavaModel, GaussianMixtureParams, JavaMLWritable, JavaMLReadable,
                           HasTrainingSummary):
```
Since currently, ```HasXXX``` has both setters and getters, so this PR adds  both the setters and getters to ```GaussianMixtureModel```.
After next refactor jira https://issues.apache.org/jira/browse/SPARK-29093 (remove automatically generated param setters in _shared_params_code_gen.py), setters will be removed from ```HasXXX```, I will need to  explicitly add ```setFeaturesCol```, ```setPredictionCol``` and ```setProbabilityCol``` to ```GaussianMixtureModel```, then the code will be as following
```
class GaussianMixtureModel(JavaModel, GaussianMixtureParams, JavaMLWritable, JavaMLReadable,
                           HasTrainingSummary):
  def setFeaturesCol
  def setPredictionCol
  def setProbabilityCol
```
It will be exactly the same as the currently scala code below: 
```
class GaussianMixtureModel extends Model with GaussianMixtureParams with MLWritable
  with HasTrainingSummary
  def setFeaturesCol
  def setPredictionCol
  def setProbabilityCol
```
I agree with you that we should retain ```@since``` annotations with the highest version of any of the removed methods.

**Comment by srowen** [comment]
OK so this will need to be followed up with another PR. That's fine, just remind me to review it. (We can link the JIRAs too to make it clear.) I'll leave it open for more comments for a bit.

**Review by zhengruifeng — APPROVED** [review]

**Comment by zhengruifeng** [comment]
One more thing:  we may need to rename `xxxParams` to `_xxxParams` in this PR & https://github.com/apache/spark/pull/25908 ?

**Comment by srowen** [comment]
PS @zhengruifeng I think you can merge this, and the other related PR, once you are both comfortable with it.

File: python/pyspark/ml/clustering.py (hunk 2 above)
Comment (zhengruifeng) [reply python/pyspark/ml/clustering.py]: This maybe a little bug, `KMeans` & `KMeansModel` should have different uid (like `LogisticRegression` & `LogisticRegressionModel`). This issue seems also happen in other place. But we can leave it here.

**Comment by zhengruifeng** [comment]
Merged to master, thanks all!

**Comment by huaxingao** [comment]
Thanks! @srowen @zhengruifeng

**Pull request closed without merging**